encoding/ssz/list.go

```diff
@@ -0,0 +1,266 @@
+func NewListVariableElementSerdes[T SerDesable](newt func() T) ListSerdes[T] {
+	return ListSerdes[T]{
+		new:       newt,
+		marshal:   MarshalListVariableElement[T],
+		unmarshal: UnmarshalListVariableElement[T],
+	}
+}
+
+// Marshal encodes a slice of elements of type [T] as an ssz-encoded List.
+func (ls ListSerdes[T]) Marshal(elems []T) ([]byte, error) {
+	if len(elems) == 0 {
+		return nil, nil
+	}
+	return ls.marshal(elems)
+}
+
+// Unmarshal decodes an ssz-encoded List of elements of type [T].
+func (ls ListSerdes[T]) Unmarshal(buf []byte) ([]T, error) {
+	if len(buf) == 0 {
+		return nil, nil
+	}
+	return ls.unmarshal(buf, ls.new)
+}
+
+// MarshalListVariableElement marshals a list of variable-sized elements.
+// A list of variable-sized elements is marshaled by first writing the offsets of each element to the
+// beginning of the byte sequence (the fixed size section of the variable sized list container), followed
+// by the encoded values of each element at the indicated offset relative to the beginning of the byte sequence.
+//
+// MarshalListFixedElement should be used for fixed-size elements.
+// SSZ Lists have different encoding rules depending whether their elements are fixed- or variable-sized,
+// and we can't differentiate them by the ssz interface, so it is the caller's responsibility to
+// pick the correct method.
+func MarshalListVariableElement[T Marshalable](elems []T) ([]byte, error) {
+	var err error
+	var total uint32
+	nElems := len(elems)
+	if nElems == 0 {
+		return nil, nil
+	}
+	sizes := make([]uint32, nElems)
+	for i, e := range elems {
+		sizes[i], err = safeUint32(e.SizeSSZ())
+		if err != nil {
+			return nil, err
+		}
+		total += sizes[i]
+	}
+	nextOffset, err := safeUint32(nElems * offsetLen)
+	if err != nil {
+		return nil, err
+	}
+	buf := make([]byte, 0, total+nextOffset)
+	for _, size := range sizes {
+		buf = binary.LittleEndian.AppendUint32(buf, nextOffset)
+		nextOffset += size
+	}
+	for _, elem := range elems {
+		buf, err = elem.MarshalSSZTo(buf)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buf, nil
+}
+
+// UnmarshalListVariableElement unmarshals a ssz-encoded list of variable-sized elements.
+// Because this generic method is parameterized by a [T Unmarshalable] interface type,
+// it is unable to initialize elements of the list internally. That is why the caller must
+// provide the `newt` function that returns a new instance of the type [T] to be unmarshaled.
+// This func will be called for each element in the list to create a new instance of [T].
+//
+// UnmarshalListFixedElement should be used for fixed-size elements.
+// SSZ Lists have different encoding rules depending whether their elements are fixed- or variable-sized,
+// and we can't differentiate them by the ssz interface, so it is the caller's responsibility to
+// pick the correct method.
+func UnmarshalListVariableElement[T Unmarshalable](buf []byte, newt func() T) ([]T, error) {
```

Comment:
Checked the logic of this function and it looks good and safe to me except the discussion offline